packages/pds/src/api/app/bsky/feed/getPostThread.ts

```diff
@@ -1,50 +1,47 @@
-import { InvalidRequestError } from '@atproto/xrpc-server'
-import * as common from '@atproto/common'
 import { Server } from '../../../../lexicon'
-import * as GetPostThread from '../../../../lexicon/types/app/bsky/feed/getPostThread'
-import DatabaseSchema from '../../../../db/database-schema'
-import { countAll } from '../../../../db/util'
-import { getDeclaration } from '../util'
-import { ImageUriBuilder } from '../../../../image/uri'
-import { embedsForPosts, FeedEmbeds } from '../util/embeds'
 import AppContext from '../../../../context'
+import {
+  ActorViewMap,
+  FeedEmbeds,
+  FeedRow,
+  FeedService,
+  PostInfoMap,
+} from '../../../../services/feed'
+import { InvalidRequestError } from '@atproto/xrpc-server'
+
+export type PostThread = {
+  post: FeedRow
+  parent?: PostThread | ParentNotFoundError
+  replies?: PostThread[]
+}
 
 export default function (server: Server, ctx: AppContext) {
   server.app.bsky.feed.getPostThread({
     auth: ctx.accessVerifier,
     handler: async ({ params, auth }) => {
       const { uri, depth = 6 } = params
       const requester = auth.credentials.did
-      const { db, imgUriBuilder } = ctx
 
-      const queryRes = await postInfoBuilder(db.db, requester)
-        .where('post.uri', '=', uri)
-        .executeTakeFirst()
+      const feedService = ctx.services.feed(ctx.db, ctx.imgUriBuilder)
 
-      if (!queryRes) {
+      const threadData = await getThreadData(feedService, uri, depth)
+      if (!threadData) {
         throw new InvalidRequestError(`Post not found: ${uri}`, 'NotFound')
       }
+      const relevant = getRelevantIds(threadData)
+      const [actors, posts, embeds] = await Promise.all([
+        feedService.getActorViews(Array.from(relevant.dids)),
+        feedService.getPostViews(Array.from(relevant.uris), requester),
+        feedService.embedsForPosts(Array.from(relevant.uris)),
+      ])
 
-      const embeds = await embedsForPosts(db.db, imgUriBuilder, [queryRes.uri])
-      const thread = rowToPost(imgUriBuilder, embeds, queryRes)
-      if (depth > 0) {
-        thread.replies = await getReplies(
-          db.db,
-          imgUriBuilder,
-          thread,
-          depth - 1,
-          requester,
-        )
-      }
-      if (queryRes.parent !== null) {
-        thread.parent = await getAncestors(
-          db.db,
-          imgUriBuilder,
-          queryRes.parent,
-          requester,
-        )
-      }
-
+      const thread = composeThread(
+        threadData,
+        feedService,
+        posts,
+        actors,
+        embeds,
+      )
```

Comment:
It's kind of a pain but I think this is the way to do it—much nicer 👍

Reply:
yeah agreed. was a bit of a pain to rewrite, but much nicer on the other side